File: TypeScript/330_Patching_Array.ts
function minPatches(nums: number[], n: number): number {
    let patches = 0;
    let k = 0; //coverage
    let lastInserted = nums[0];
    let lastIx = -1;
    while (k<n) {
        let s = k+1;
        let ix = nums.findIndex((num,i) => i>lastIx && num >= lastInserted && num<= s);
        if(ix != -1) {
            let t = nums[ix];
            console.log({ix,t,s, lastInserted, nums,k,newK: k+t})
            k = k + t;
            lastInserted = t;
            lastIx=ix;
            continue;
        }
        ix = binarySearch(nums, s);
        if(ix < 0){
            patches++;
            nums.splice(-(ix+1), 0, s); 
            lastInserted=s;
            lastIx = -(ix+1)
        } else {
            lastIx = ix;
        }
        k = k + s;
    }
    console.log({ nums, k })
    return patches;
};

function binarySearch(arr, x) {

    let start = 0, end = arr.length - 1;

    // Iterate while start not meets end
    while (start <= end) {

        // Find the mid index
        let mid = Math.floor((start + end) / 2);

        // If element is present at 
        // mid, return True
        if (arr[mid] === x) return mid;

        // Else look in left or 
        // right half accordingly
        else if (arr[mid] < x)
            start = mid + 1;
        else
            end = mid - 1;
    }

    return (-(start) -1);
}

File: Python/330_Patching_Array.py
class Solution:
    def minPatches(self, nums: List[int], n: int) -> int:
        patches = 0
        k = 0
        lastInserted = nums[0]
        lastIx = -1
        while (k<n):
            s = k+1
            # print({'lower':lastInserted, 'upper': s, 'nums': nums, 'lastIx': lastIx})
            ixs = [i for i in range(len(nums)) if i > lastIx and  nums[i] >= lastInserted and nums[i] <= s]
            if (len(ixs) > 0):
                ix = ixs[0]
                t=nums[ix]
                # print(ixs)
                # print({ 'ix': ix, 't': t, 'k': k, 's': s})
                k = k +t
                lastInserted=t
                lastIx=ix
                continue

            ix = self.binarySearch(nums,s)
            # print({'ix':ix, 's': s})
            if(ix<0):
                patches = patches + 1
                nums.insert(-(ix+1),s)
                lastInserted = s
                lastIx = -(ix +1)
            else:
                lastIx = ix
            
            k = k+s
        
        return patches


    def binarySearch(self, arr: List[int], n: int):
        start=0
        end=len(arr)-1
        while start <= end:
            mid = math.floor((start+end) /2 )

            if(arr[mid] == n): return mid
            elif(arr[mid] < n ): start = mid+1
            else: end = mid-1

        return (-(start)-1)


        
